commands/core/site_install.drush.inc

```diff
@@ -181,7 +187,12 @@ function drush_core_pre_site_install($profile = NULL) {
     drush_bootstrap(DRUSH_BOOTSTRAP_DRUPAL_SITE);
   }
 
-  $sql->drop_or_create($db_spec);
+  if ($db_drop_existing_tables) {
+    $sql->drop_or_create();
```

Comment:
I think we should remove  drop_or_create() method entirely. It is used in archive-restore only. Just copy its code to there.